plugins/__init__.py

```diff
@@ -251,6 +305,157 @@ def serialize_config(config_block: "terraform::config::Block") -> "dict":  # typ
+        # as our config block in the global config
+        if not isinstance(state_container, SequenceProxy):
+            raise PluginException(
+                f"The state dict has an unexpected value at key {config.name}: "
+                f"{state_container} ({type(state_container)}).  State dict is "
+                f"{parent_state} ({type(parent_state)})"
+            )
+
+        # This is the key of our config
+        config_key = config.key
+
+        parent_config = config.parent
+        sibling_configs = parent_config.children
+
+        # These are all the config keys, sorted (the state should be sorted the same way)
+        config_keys = sorted(c.key for c in sibling_configs if c.name == config.name)
+
+        # Sanity check, the state_container should have the same length as our configs
+        if not len(state_container) == len(config_keys):
+            raise PluginException(
+                "The length of the state list doesn't match the number of config blocks there is.  "
+                f"state={state_container}, config_keys={config_keys}"
+            )
+
+        config_position = config_keys.index(config_key)
+        return state_container[config_position]
+
+    if config.nesting_mode == "set":
+        # Block embedded in a set, the matching state might be anywhere.  To find it,
+        # we rely on the fact that the config dict is a subset of the state dict.  We
+        # check for each element in the list, which one wouldn't we changed it we
+        # updated it with our config, that one should be our state.
+        if not isinstance(state_container, SequenceProxy):
+            raise PluginException(
+                f"The state dict has an unexpected value at key {config.name}: "
+                f"{state_container} ({type(state_container)}).  State dict is "
+                f"{parent_state} ({type(parent_state)})"
+            )
+
+        # This is our config dict, minus all the default (null) values
+        clean_config = {k: v for k, v in config._config.items() if v is not None}
+
+        matching_states: list[DictProxy] = []
+        for candidate_state in state_container:
+            state = {k: v for k, v in candidate_state.items()}
+            for key, value in clean_config.items():
+                state[key] = value
+
+            if state == candidate_state:
+                matching_states.append(candidate_state)
+
+        if len(matching_states) != 1:
+            raise PluginException(
+                f"Failed to find a unique matching state in the list {state_container} for config "
+                f"{clean_config}.  Got a total of {len(matching_states)} in {matching_states}"
+            )
+
+        return matching_states[0]
+
+    raise PluginException(f"Unknown nesting mode: {config.nesting_mode}")
+
+
+@plugin
+def get_from_unknown_dict(object: "dict", key: "string") -> "any":  # type: ignore
```

Comment:
isn't that what the normal '[ ]' operator does?

Reply:
Support for this is a recent addition (it wasn't there when I started working on this): https://github.com/inmanta/inmanta-core/commit/30b491d039a1fb6289f3f67846f20ca5475872fb
But I guess that given that we don't use it in this model (this was going to be used by the generated modules), we can leave it out, and add it depending on the minimal compiler version.